Mod bufferobject

hjunqq opened the pull request:
<!--
👋 Hello, and thank you for starting this contribution!
📖 Make sure you've read our CONTRIBUTING.md guide before submitting your pull request.
❗️ Please follow the template below to help other contributors review your work.
-->

### Context
<!--
Explain why this change is needed. Please include relevant links supporting this change, such as:
- fix #ISSUE_NUMBER (from issue tracker)
- discourse post thread, or any other existing references
- screenshot of the issue
- console log of error, callstack
-->

fix bug
```bash
TS7016: Could not find a declaration file for module BufferObject and Texture
```


### Results
<!--
Describe or illustrate the effects of your contribution. Please include:
- comparisons of the behavior before vs after
- screenshots of new or changed visualizations if applicable
-->

No Errors 

### Changes
<!--
Please describe what is changing. Include:
- APIs added, deleted, deprecated, or changed
- Classes and methods added, deleted, deprecated, or changed
- A summary of usage if this is a new feature or change to an API. Adequate documentation and TS definitions should also be added/updated.
-->
- [x] Documentation and TypeScript definitions were updated to match those changes
- [x] TypeScript definitions were updated

### PR and Code Checklist
<!--
NOTE: We will not merge if the following steps have not been completed!
-->
- [x] [semantic-release](https://github.com/semantic-release/semantic-release) commit messages
- [x] Run `npm run reformat` to have correctly formatted code

### Testing
<!--
Please describe how this can be tested by reviewers. Be specific about anything not tested and the reasons why.
Tests should complete without errors. See CONTRIBUTING.md
-->
- [x] This change adds or fixes unit tests <!-- Tests should be added for new functionality -->
- [x] Tested environment:
  - **vtk.js**: <!-- ex: 14.0.0 (favor latest master) -->
  - **OS**: <!-- ex: Windows 10, iOS 13.6 -->
  - **Browser**: <!-- ex: Chrome 89.0.4389.128 -->

<!--
Edit and uncomment the section below if relevant

### Funding
This contribution is funded by [Example](https://example.com).

 -->

floryst commented:
Hi, thanks for the contribution!

Your changes define interfaces using the default values rather than the actual types. I would recommend looking at the code for `vtkOpenGLTexture` and other PRs (e.g. https://github.com/Kitware/vtk-js/pull/2976) for examples on how to add new type definitions.

hjunqq commented:
I have made a series of enhancements to **vtkOpenGLTexture** and   **vtkOpenGLBufferObject** to improve its type safety and readability. Here is an overview of the key changes:

Consistency and Clarity in Type Definitions: Following the precedent set by previous Pull Requests, such as  #2976 , I ensured that my type definitions are consistent with the existing code in the project. I introduced new types to more clearly represent some complex structures and concepts.

Enhanced Documentation and Comments: I have added detailed comments and documentation, explaining the new types and the modified sections. This will help other developers understand the purpose of these changes and how to use them.

I believe these changes will enhance the robustness and maintainability of **vtkOpenGLTexture**  and   **vtkOpenGLBufferObject**  and look forward to everyone's feedback. Thank you for your time and consideration.

floryst commented on Sources/Common/DataModel/SelectionNode/index.d.ts:
```diff
@@ -54,7 +54,7 @@ export interface vtkSelectionNode extends vtkObject {
 	 * This functions is called internally by VTK.js and is not intended for public use.
 	 */
-	setProperties(properties: ISelectionNodeProperties);
+	setProperties(properties: ISelectionNodeProperties):any;
```
```suggestion
	setProperties(properties: ISelectionNodeProperties): boolean;
```

This is an auto-generated setter, so it returns boolean. (Returns true if the underlying value is actually changed).

floryst commented on Sources/Common/DataModel/SelectionNode/index.d.ts:
```diff
@@ -64,7 +64,7 @@ export interface vtkSelectionNode extends vtkObject {
 	 * This functions is called internally by VTK.js and is not intended for public use.
 	 */
-	setSelectionList(selectionAttributeIDs: ISelectionNodeProperties);
+	setSelectionList(selectionAttributeIDs: ISelectionNodeProperties):any;
```
Same thing here: should be `boolean`

floryst commented on Sources/Rendering/OpenGL/BufferObject/index.d.ts:
```diff
@@ -0,0 +1,87 @@
+ * Interface for initial values of BufferObject
+ */
+export interface IBufferObjectInitialValues {
+  objectType: ObjectType;
+  context?: WebGLRenderingContext | WebGL2RenderingContext;
+  allocatedGPUMemoryInBytes: number;
```
`objectType` and `allocatedGPUMemoryInBytes` should be optional, since generally initial values are optional.

floryst commented on Sources/Rendering/OpenGL/BufferObject/index.d.ts:
```diff
@@ -0,0 +1,87 @@
+/**
+ * Base type for vtkOpenGLBufferObject excluding the 'set' method from vtkObject
+ */
+type vtkOpenGLBufferObjectBase = Omit<vtkObject, 'set'> & vtkAlgorithm;
+
+/**
+ * Interface for OpenGL Buffer Object
+ */
+export interface vtkOpenGLBufferObject extends vtkOpenGLBufferObjectBase {
```
It looks to me that `vtkOpenGLBufferObject` merely extends `vtkObject` instead of what you have defined here.

floryst commented on Sources/Rendering/OpenGL/Texture/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+/**
+ * Initial values for creating a new instance of vtkOpenGLTexture.
+ */
+export interface ITextureInitialValues {
```
All fields should be marked as optional.

floryst commented on Sources/Rendering/OpenGL/Texture/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+  _forceInternalFormat: boolean;
+  context?: WebGLRenderingContext | WebGL2RenderingContext;
+  handle: number;
+  sendParametersTime: Time;
+  textureBuildTime: Time;
```
These should be typed as `vtkObject`, since `macro.obj` creates a base VTK object.

floryst commented on Sources/Rendering/OpenGL/Texture/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+import { vtkAlgorithm, vtkObject } from "../../../interfaces";
+import { VtkDataTypes } from "../../../Common/Core/DataArray";
+
+/**
+ * Represents the time with a modification timestamp.
+ */
+interface Time{
+  mtime:number;
+}
```
You can drop this interface; see my comment below as to the correct typing.

floryst commented on Sources/Rendering/OpenGL/Texture/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+  height: number;
+  depth: number;
+  autoParameters: boolean;
+  wrapS: Wrap.CLAMP_TO_EDGE;
+  wrapT: Wrap.CLAMP_TO_EDGE;
+  wrapR: Wrap.CLAMP_TO_EDGE;
+  minificationFilter: Filter.NEAREST;
+  magnificationFilter: Filter.NEAREST;
```
These should be the enum classes, not enum values.

floryst commented on Sources/Rendering/OpenGL/Texture/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+  maxLevel: number;
+  generateMipmap: boolean;
+  useHalfFloat: boolean;
+  oglNorm16Ext: null | undefined;
```
Should be typed as `any`, since I can't find a suitable type for the EXT_texture_norm16 extension.

floryst commented on Sources/Rendering/OpenGL/Texture/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+/**
+ * Interface for OpenGL Texture.
+ */
+export interface vtkOpenGLTexture extends vtkOpenGLTextureBase {
```
This should be `vtkOpenGLTexture extends vtkViewNode`, based on what vtkOpenGLTexture extends in the codebase.

floryst commented on Sources/Rendering/OpenGL/Texture/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+   * Renders the texture within the given render window.
+   * @param renWin The render window in which to render the texture.
+   */
+  render(renWin:any):void;
```
```suggestion
  render(renWin: vtkOpenGLRenderWindow):void;
```

floryst commented on Sources/Rendering/OpenGL/Texture/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+   * Releases the graphics resources used by the texture within the given render window.
+   * @param renWin The render window whose resources should be released.
+   */
+  releaseGraphicsResources(renWin:any):void;
```
```suggestion
  releaseGraphicsResources(renWin: vtkOpenGLRenderWindow):void;
```

floryst commented on Sources/Rendering/OpenGL/Texture/index.d.ts:
```diff
@@ -0,0 +1,316 @@
+   * Sets the OpenGL render window in which the texture will be used.
+   * @param renWin The render window to set.
+   */
+  setOpenGLRenderWindow(renWin:any):void;
```
```suggestion
  setOpenGLRenderWindow(renWin: vtkOpenGLRenderWindow):void;
```

floryst commented on Sources/Widgets/Core/AbstractWidgetFactory/index.d.ts:
```diff
@@ -72,7 +72,7 @@ export interface vtkAbstractWidgetFactory extends vtkObject {
    * @param {Bounds} bounds 
    */
-  placeWidget(bounds: Bounds);
+  placeWidget(bounds: Bounds):any;
```
This should be `void` instead of `any`.

floryst reviewed (commented):
Thanks for the update! I think it should be good to go after some changes.

There are some whitespace quirks that would be nice to adjust, but I'll just have prettier enforce formatting on `d.ts` files in a future PR.

hjunqq commented:
Thank you for your feedback. I have made the necessary corrections as per your suggestions and also adjusted the spacing issues.

floryst commented on Sources/Rendering/OpenGL/BufferObject/index.d.ts:
```diff
@@ -0,0 +1,82 @@
+ * Interface for initial values of BufferObject
+ */
+export interface IBufferObjectInitialValues {
+  objectType?: ObjectType | ObjectType.ARRAY_BUFFER;
+  context?: WebGLRenderingContext | WebGL2RenderingContext;
+  allocatedGPUMemoryInBytes?: number | 0;
```
```suggestion
  objectType?: ObjectType;
  context?: WebGLRenderingContext | WebGL2RenderingContext;
  allocatedGPUMemoryInBytes?: number;
```

hjunqq replied:
Modified the above problem

floryst commented:
This is good to go after my most recent comment.
